Studio-Lovelies/GG-JointJustice-Unity
# Automatic documentation generation

## Summary
This branch includes `/docs_generator`, which generates [`.MD`/Markdown-files](https://daringfireball.net/projects/markdown/) for the `ActionDecoder.cs`, identical to the current structure of our wiki.  
These are generated for every push and PR to `develop`, `releasecandidate/*` and `master` branch using GitHub Actions.

- Actor Actions ([wiki](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/wiki/Actor-Actions) vs. [generated](https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/docs/methods/Actor.md))
- Preset Values ([wiki](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/wiki/Preset-Values) vs. [generated](https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/docs/constants.md))

### Technical implementation
`/docs_generator` contains a .NET 6 C# Console App, which reads [`Assets/Scripts/TextDecoder/ActionDecoder.cs`](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/f14af4e2bba25c64cb981eb07d41b2392de26632/Assets/Scripts/TextDecoder/ActionDecoder.cs) and generates markdown files based on the [xmldoc comments](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/) of `CAPITALZED_METHODS()`.
Files are generated for...
-  **every category** methods are assigned to, e.g.:
  https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/unity/Assets/Scripts/TextDecoder/ActionDecoder.cs#L151
  https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/unity/Assets/Scripts/TextDecoder/ActionDecoder.cs#L293

- and a `constants.md` containing **all files for `AssetNames`** (like sound effects, background images or actors), e.g.:
  https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/unity/Assets/Scripts/TextDecoder/ActionDecoder.cs#L222
  https://github.com/ViMaSter/GG-JointJustice-Unity/blob/develop/unity/Assets/Scripts/TextDecoder/ActionDecoder.cs#L517

## Minor notices
- `CLEAR_SPEED()`, `WAIT_FOR_INPUT()`, `NARRATE()` exist in the Wiki, but don't exist in code.
- `RELOAD_SCENE()` has no Wiki entry.
- The actor file for for `JudgeBrent` was called `BrentJudge`. Updated it to `JudgeBrent`, to be consistent.
- **Qualified `AssetName`s even more explicitly**:
  - [The wiki has a page on "preset values"](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/wiki/Preset-Values), which lists which characters we have, their animations, our music, etc.
    In order for the generator to somehow figure out, which files are applicable where, I've split up the `AssetName` classes:
    ```cs
    /// <param name="evidence" validFiles="Assets/ScriptableObjects/Evidence/*.asset">Name of evidence</param>
    private void REMOVE_EVIDENCE(EvidenceAssetName evidence) { }
    ```
    In this case, `EvidenceAssetName` is mapped to all files matching the `Assets/ScriptableObjects/Evidence/*.asset` qualifier
    The new classes are:
     - FullscreenAnimationAssetName
     - SceneAssetName
     - SfxAssetName
     - SongAssetName
     - EvidenceAssetName
     - ActorPoseAssetName
     - ActorAssetName

## Before/after screenshots and/or animated gif
# ⚠️ MERGING THIS MOVES THE GAME! `/` -> `/unity`! ⚠️ 
This is how our repo will look after this merge: https://github.com/ViMaSter/GG-JointJustice-Unity
All GitHub Actions are updated accordingly. Also the game build is only triggered, if files inside this `unity` subfolder change.

The folders:
- `/unity`: contains everything related to the Unity project
- `/docs_generator`: contains a .NET 6 Console Application which generates `/docs`
- `/docs`: contains [`.MD`/Markdown-files](https://daringfireball.net/projects/markdown/) which are generated whenever the project inside `/docs_generator` is run  
  (currently on every PR to `develop`, `releasecandidate/*` and `master`)


## Testing instructions
Pull down this PR and make sure everything inside Unity still works.

After switching to this branch, your working directory might still contain previously untracked files:
![image](https://user-images.githubusercontent.com/1689033/148837753-6eee9957-be98-4072-977d-6ba59b29329d.png)
Either...
- move them to `/unity` by hand (your Unity project should load them as normal)
- run `git clean -d -n` (this will **not** remove any files yet) to see which files are unused and **then** run `git clean -d -f`  
  (❗ this method **will delete** the previously listed files and Unity will have to re-import all files, so your first project startup might take a little longer)

## Additional information

- `[ ]` Changes UI
- `[ ]` Introduces new feature
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:
